File: Examples/Comparison/fastq_compare.jl
pushfirst!(LOAD_PATH, abspath(joinpath(@__DIR__, "..", "..")))

using BioToolkit
include(joinpath(@__DIR__, "..", "..", "scripts", "benchmark_helpers.jl"))

function write_sample_fastq(path::AbstractString, record_count::Integer, read_length::Integer)
    bases = ('A', 'C', 'G', 'T')

    open(path, "w") do io
        for record_index in 1:record_count
            write(io, "@read$(record_index)\n")

            for base_index in 1:read_length
                base = bases[mod(record_index + base_index - 2, length(bases)) + 1]
                write(io, base)
            end
            write(io, "\n")

            write(io, "+\n")
            for _ in 1:read_length
                write(io, 'I')
            end
            write(io, "\n")
        end
    end
end

function main()
    record_count = 30_000
    read_length = 100

    mktempdir() do dir
        fastq_path = joinpath(dir, "sample.fastq")
        python_script = joinpath(@__DIR__, "fastq_compare.py")
        written_fastq_path = joinpath(dir, "written.fastq")

        write_sample_fastq(fastq_path, record_count, read_length)

        BioToolkit.read_fastq(fastq_path)

        julia_ms, julia_records = elapsed_ms() do
            BioToolkit.read_fastq(fastq_path)
        end

        write_ms, _ = elapsed_ms() do
            BioToolkit.write_fastq(written_fastq_path, julia_records)
        end

        python_output = read(`conda run -n general --no-capture-output python $python_script $fastq_path`, String)

        total_bases = sum(length(record.sequence) for record in julia_records)

        println("FASTQ parse/write benchmark")
        println("  records: ", record_count)
        println("  read_length: ", read_length)
        println("  julia_fastq_ms=", round(julia_ms, digits=4))
        println("  julia_fastq_records=", length(julia_records))
        println("  julia_fastq_bases=", total_bases)
        println("  julia_fastq_write_ms=", round(write_ms, digits=4))
        print(python_output)
    end
end

main()

File: Examples/Comparison/fastq_compare.py
from sys import argv
from time import perf_counter

from Bio import SeqIO


def time_call(func):
    start = perf_counter()
    result = func()
    return (perf_counter() - start) * 1000.0, result


def parse_fastq(fastq_path):
    with open(fastq_path) as handle:
        return list(SeqIO.parse(handle, "fastq"))


def write_fastq(records, output_path):
    with open(output_path, "w") as handle:
        return SeqIO.write(records, handle, "fastq")


def main():
    if len(argv) != 2:
        raise SystemExit("usage: fastq_compare.py <fastq-path>")
    fastq_path = argv[1]
    parse_fastq(fastq_path)
    py_ms, py_records = time_call(lambda: parse_fastq(fastq_path))
    output_path = fastq_path + ".written"
    write_fastq(py_records, output_path)
    py_write_ms, py_written = time_call(lambda: write_fastq(py_records, output_path))
    print(f"python_fastq_ms={py_ms:.4f}")
    print(f"python_fastq_records={len(py_records)}")
    print(f"python_fastq_bases={sum(len(record.seq) for record in py_records)}")
    print(f"python_fastq_write_ms={py_write_ms:.4f}")
    print(f"python_fastq_write_records={py_written}")


if __name__ == "__main__":
    main()
